llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp

```diff
@@ -2591,15 +2590,31 @@ static Instruction *foldCastShuffle(ShuffleVectorInst &Shuf,
     return nullptr;
   }
 
+  VectorType *CastSrcTy = cast<VectorType>(Cast0->getSrcTy());
   VectorType *ShufTy = Shuf.getType();
   VectorType *ShufOpTy = cast<VectorType>(Shuf.getOperand(0)->getType());
-  VectorType *CastSrcTy = cast<VectorType>(Cast0->getSrcTy());
 
   // TODO: Allow length-increasing shuffles?
   if (ShufTy->getElementCount().getKnownMinValue() >
       ShufOpTy->getElementCount().getKnownMinValue())
     return nullptr;
 
+  // shuffle (cast X), Poison, identity-with-extract-mask -->
+  // cast (shuffle X, Poison, identity-with-extract-mask).
+  if (Cast0->hasOneUse() && Shuf.isIdentityWithExtract() &&
+      isa<PoisonValue>(Shuf.getOperand(1))) {
```

Comment:
```suggestion
  if (isa<PoisonValue>(Shuf.getOperand(1) && 
       Cast0->hasOneUse() && Shuf.isIdentityWithExtract())) {
```

Do you really need to check for poison? I would assume that's covered under the Shuf check?

Reply:
Thanks, updated the check. I think the poison check is needed, because `isIdentityWithExtract` also allows to select all elements from the second vector operand, as @dtcxzyw 's test case showed